[#6636] feat(client-py): add Python 3.12 support for client testing

AndreVale69 opened the pull request:
### What changes were proposed in this pull request?

Tests could not be performed with Python 3.12, although the client supports version 3.12.

### Why are the changes needed?

It's important to test the Python client with all supported versions.

Fix: #6636

### Does this PR introduce _any_ user-facing change?

No at all. It only updates some dependecies:
- Patch:
  - `requests` from `2.32.2` to [`2.32.3`](https://github.com/psf/requests/releases/tag/v2.32.3) (latest)
  - `dataclasses-json` from `0.6.6` to [`0.6.7`](https://github.com/lidatong/dataclasses-json/releases/tag/v0.6.7) (latest)
- Minor:
  - `llama-index` from `0.10.40` to [`0.11.18`](https://github.com/run-llama/llama_index/blob/main/CHANGELOG.md#llama-index-core-01118) (this version supports python 3.13)
  - `cachetools` from `5.3.3` to [`5.5.2`](https://github.com/tkem/cachetools/blob/master/CHANGELOG.rst#v552-2025-02-20) (latest)
  - `pandas` from `2.0.3` to [`2.2.3`](https://github.com/pandas-dev/pandas/releases/tag/v2.2.3) (if py version > 3.8)

### How was this patch tested?

[how-to-use-python-client.md](https://github.com/apache/gravitino/blob/main/docs/how-to-use-python-client.md)

### Note

I also tried Python 3.13, but it's not supported by `pandas`. See: https://github.com/pandas-dev/pandas/issues/59057

AndreVale69 commented:
> How do you know that we don't support Python 3.12?

I've been trying to install [requirements-dev.txt](https://github.com/apache/gravitino/blob/main/clients/client-python/requirements-dev.txt) using pip on python 3.12 and I've noticed some errors during the pandas installation.

So I did some research and found that [pandas has released support for python 3.12 on November 2023](https://github.com/pandas-dev/pandas/issues/53665#issuecomment-1797119460), but the version 2.0.3 required by [requirements-dev.txt](https://github.com/apache/gravitino/blob/main/clients/client-python/requirements-dev.txt) was released on [23 June 2023](https://github.com/pandas-dev/pandas/releases/tag/v2.0.3).

Pandas supports py 3.12 from the version 2.2.0:
- [Installation documentation of 2.1.4](https://pandas.pydata.org/pandas-docs/version/2.1/getting_started/install.html)
  
  ![2-1-4](https://github.com/user-attachments/assets/d80834d4-a736-45d5-a3b1-4c117823631c)

- [Installation documentation of 2.2.0](https://pandas.pydata.org/pandas-docs/version/2.2/getting_started/install.html)
  
  ![2-2](https://github.com/user-attachments/assets/d98d897a-fcaf-456f-a949-bbde3ec2cc70)

AndreVale69 commented:
> I see, thanks. Gravitino python client alone should work with Python 3.12, I used it locally. Pandas is only used for test/dev, so I think the release package should be worked with 3.12, the dev dependency cannot work with 3.12. Maybe we should make it work with 3.12, what do you think?

Oh, good point, you are right. I can try playing with the dependencies to see if everything works with python 3.12. However, pandas 2.2 no longer supports python 3.8 (and it's no longer supported by python). So what should we do?

jerryshao commented:
@AndreVale69 I just check the EOL of Python 3.8, looks like Python 3.8 has already been EOL, so I think we can remove the support of Python 3.8, move on to support 3.12, what do you think?

AndreVale69 commented:
@jerryshao I just fixed it and released it! I'm not dropping the 3.8 version because the released client uses it. So to test it, we need to have python 3.8 support. For this reason, I add a line in the requirements to install the correct pandas version depending on the python that runs.

I've also updated the CI and the documentation. Now testing with Python 3.12 works! It is very important to have all possible versions and to test the client, otherwise there might be some problems. Now we can do that.

Since I've been working with dependecies, I've also bumped some versions to the latest (some fixes). In commit d897c73 I've also added the changelos.

jerryshao commented:
Thanks, I will review it soon.

jerryshao commented:
Can you please update the PR title and description, since it is not valid anymore?

jerryshao commented on clients/client-python/setup.py:
```diff
@@ -53,7 +53,7 @@
         "Programming Language :: Python :: 3.10",
         "Programming Language :: Python :: 3.11",
-        "Programming Language :: Python :: 3.12",
+        'Programming Language :: Python :: 3.12',
```
Any reason to change to `'` instead of `"`?

AndreVale69 replied:
My bad, I am fixing it right now.

jerryshao commented:
I think you should also update the PR description @AndreVale69 .

AndreVale69 commented:
> I think you should also update the PR description @AndreVale69 .

Yes, I'm writing it.

AndreVale69 commented:
@jerryshao PR changed and typo fixed. Thanks for your patience.

jerryshao reviewed (approved):
LGTM. @AndreVale69 thanks for the PR.
